scipy/optimize/_bracket.py

```diff
@@ -179,9 +177,19 @@ def _bracket_root(func, xl0, xr0=None, *, xmin=None, xmax=None, factor=None,
     # We don't need to retain the corresponding function value, since the
     # fixed end of the bracket is only needed to compute the new value of the
     # moving end; it is never returned.
+    xmin = np.broadcast_to(xmin, shape).astype(dtype, copy=False)
+    xmax = np.broadcast_to(xmax, shape).astype(dtype, copy=False)
 
-    xmin = np.broadcast_to(xmin, shape).astype(dtype, copy=False).ravel()
-    xmax = np.broadcast_to(xmax, shape).astype(dtype, copy=False).ravel()
+    status = np.full_like(x, eim._EINPROGRESS, dtype=int)  # in progress
+    # Stop with error code when initial bracket is invalid.
+    i = ~((xmin <= xl0) & (xl0 < xr0) & (xr0 <= xmax))
+    if i.ndim == 0:
+        i = np.array([i, i])
+    else:
+        i = np.concatenate((i, i))
```

Comment:
Let me know if you can think of a cleaner way to do this. 

To explain what's going on here:

I try generate a boolean index `i` into `status` for the places where the initial bracket is invalid.

* It seems like it's only after `broadcast_to`ing `xmin` and `xmax`, but before raveling them, that `xmin`, `xl0`, `xr0`, and `xmax` have compatible shapes, so that's where I generate the boolean index. 
* The boolean index `i` needs to be concatenated with itself to index into the array that got doubled to handle both left and right searches. 
* You can't concatenate 0d arrays so that has to be a special case. 
* After that we can ravel `xmin`, `xmax`, and the boolean index `i`.

After this we can use the boolean index to set the appropriate error code signaling an invalid initial bracket. `_check_terminate` checks if this error code is set, and if so, terminates those cases.

Reply:
> It seems like it's only after broadcast_toing `xmin` and `xmax`, but before raveling them, that `xmin`, `xl0`, `xr0`, and `xmax` have compatible shapes, so that's where I generate the boolean index.

I see. That is because `xl0` and `xr0` can additionally get broadcasted against each of the `args` inside `eim._initialize`, so the shapes will be *compatible* with `xmin`/`xmax` broadcasted to shapes, but they won't necessarily have the same size. 

I think it will look a little cleaner if you move this logic up between the call to `eim._initialize` and `np.concatenate(xs)`. I don't think there's anything between there and here that you need, and it will be a bit easier to think through the shapes without extra code in between.

Here's the key: after `xs` gets unpacked from `temp`, it is a 1D array. The first half (of length `n = len(xs[0])`) is `xl0` broadcasted to `shape` and `ravel`ed, and the second half is `x0r` broadcasted to `shape` and `ravel`ed. I think these halves are exactly what you need. Extract them into variables (`xl0` and `xr0`, if you please, since those names don't appear to be used anymore, and that is what these are). Then, you can broadcast `xmin` and `xmax` to `shape`, `ravel` them, and do your comparisons to get `i`. Now `i` guaranteed to be `1D` so you don't need special casing, and you can do the concatenation alongside the others.

Reply:
That works. Thanks!